# QC'd - Quality control and Contamination Detection 


## Synopsis
The pipeline calculates sequencing coverage and read depth, assess read quality and screen data against a reference database to detect contamination.

## Input

- Filenames of forward paired/single-end reads. One filename per line
- Minikraken or custom pre-built Kraken database
- Type of analysis to run. Options: coverage,quality,kraken_contamination,kraken_report,coverage_depth.

```

usage: pipeline.py [-h] [-samples SAMPLES] [-config CONFIG] [-dir DIRECTORY]
                   [-analysis ANALYSIS_NAMES] [-o OUTPUT_FOLDER] [-type TYPE]
                   [-cluster CLUSTER] [-genome_size SIZE] [-prefix PREFIX]
                   [-reference REFERENCE]

QC'd - Quality control and Contamination Detection

optional arguments:
  -h, --help            show this help message and exit

Required arguments:
  -samples SAMPLES      Filenames of forward-paired end or single-end reads. One filename per line
  -dir DIRECTORY        Path to Sequencing Reads Data directory. NOTE: Provide full/absolute path.
  -analysis ANALYSIS_NAMES
                        comma-seperated analysis names to run
                        [Options: coverage,quality,kraken_contamination,kraken_report,coverage_depth].
                         Example: -analysis coverage,quality
  -o OUTPUT_FOLDER      Output Folder Path ending with output directory name to save the results.
                        Creates a new output directory path if it doesn't exist.
                        NOTE: Provide full/absolute path.
  -type TYPE            Type of analysis: SE or PE
  -genome_size SIZE     Estimated Genome Size
  -prefix PREFIX        Prefix to use to save results files

Optional arguments:
  -config CONFIG        Path to Config file, Make sure to check config settings before running pipeline.
                        Set Karken database path under [kraken] config section
  -cluster CLUSTER      Run in one of the two modes. Default: local.
                        The possible modes are: parallel-local/local
                        parallel-local: Run jobs for each sample in parallel but on local system.
  -reference REFERENCE  Reference genome to be uses for calculating GATK Depth of coverage.

```

**Analysis options:**


> **coverage:** calculate raw sequencing coverage given a genome size; Assumes all filename/samples belong to one species type.

> **quality:** run FastQC and generate quality reports. Also, merge multiple fastqc reports to generate MultiQC reports. 

> **kraken_contamination:** Scan reads against a minikraken or pre-built custom kraken database to estimate species abundance.

> **kraken_report:** Generate user-friendly Kraken and Krona html reports for visualization.

> **coverage_depth:** calculate read depth using GATK DepthofCoverage tool. Requires a reference genome for read mapping.

**Results:**

> 1. coverage: tab-seperated prefix_Final_Coverage.txt report.

> 2. quality: FastQC/MultiQC html reports will be generated under prefix_Fastqc

> 3. kraken_contamination: Results generated by Kraken will be saved under prefix_kraken.  

> 4. kraken_report: A summary report prefix_Kraken_report_final.csv will be generated using kraken results.

> 5. coverage_depth: GATK depth_of_coverage statistics for each sample against the reference genome and a combined summary report prefix_Final_Coverage_depth.txt.

