# QC'd - Quality control and Contamination Detection 


## Synopsis

Run various QC steps on the raw sequencing data. 

1. Calculate raw sequencing coverage.
2. Run fastQC and generate a multiQC report to assess the quality of samples.
3. Screen samples against a Kraken database and detect if the data contains the expected target species.
4. Calculate read depth from BWA mapped alignments.
5. Run ARIBA MLST typing and type the samples.
6. Detect presence of antibiotic resistance genes using ARIBA CARD.

## Input

- A file containing filenames of forward paired/single-end reads. One filename per line.
- Path to Minikraken or custom pre-built Kraken database.
- Type of analysis to run. Options: coverage,quality,kraken_contamination,kraken_report,coverage_depth,mlst.

## Analysis options:


**coverage:** calculate raw sequencing coverage given a genome size; Assumes all filename/samples belong to one species type.

**quality:** run FastQC and generate quality reports. Also, merge multiple fastqc reports to generate MultiQC reports. 

**kraken_contamination:** Scan reads against a minikraken or pre-built custom kraken database to estimate species abundance.

**kraken_report:** Generate user-friendly Kraken and Krona html reports for visualization.

**coverage_depth:** calculate read depth using GATK DepthofCoverage tool. Requires a reference genome for read mapping.

**mlst:** characterize samples with ARIBA MLST typing

**amr:** detect antibiotic resistance genes using ARIBA and CARD database.


```

usage: qc.py [-h] [-samples SAMPLES] [-config CONFIG] [-dir DIRECTORY]
             [-analysis ANALYSIS_NAMES] [-o OUTPUT_FOLDER] [-type TYPE]
             [-cluster CLUSTER] [-genome_size SIZE] [-prefix PREFIX]
             [-reference REFERENCE]

QC'd - Quality control and Contamination Detection

optional arguments:
  -h, --help            show this help message and exit

Required arguments:
  -samples SAMPLES      A file containing filenames of forward-paired end or single-end reads. One filename per line
  -dir DIRECTORY        Absolute/full path to Sequencing Reads Data directory.
  -analysis ANALYSIS_NAMES
                        comma-seperated analysis names to run
                        [Options: coverage,quality,kraken_contamination,kraken_report,coverage_depth].
                        Example: -analysis coverage,quality
  -o OUTPUT_FOLDER      Output Folder Path ending with output directory name to save the results.
                        Creates a new output directory path if it doesn't exist.
  -type TYPE            Type of analysis: SE or PE
  -genome_size SIZE     Estimated Genome Size
  -prefix PREFIX        Prefix to use to save result files

Optional arguments:
  -config CONFIG        Path to Config file, Make sure to check config settings before running pipeline.
                        Set Kraken database path under [kraken] config section
  -cluster CLUSTER      Run in one of the two modes. Default: local.
                        Modes: parallel-local/local
                        parallel-local: Run jobs for each sample in parallel but on local system.
  -reference REFERENCE  Reference genome to use for calculating GATK Depth of coverage.

```

## Results:

**coverage:** tab-seperated prefix_Final_Coverage.txt report.

**quality:** FastQC/MultiQC html reports will be generated under prefix_Fastqc

**kraken_contamination:** Results generated by Kraken will be saved under prefix_kraken.  

**kraken_report:** A summary report prefix_Kraken_report_final.csv will be generated using kraken results.

**coverage_depth:** GATK depth_of_coverage statistics for each sample against the reference genome and a combined summary report prefix_Final_Coverage_depth.txt.

